In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

In [2]:
df = pd.read_csv('../data/dseb63_application_train.csv', index_col=0)

In [3]:
bureau = pd.read_csv('../data/dseb63_bureau_general_v2.csv', index_col=0)
bureau_columns = bureau.columns
# bureau.drop(columns=['SK_ID_BUREAU'], inplace=True)

# new_bureau_columns = {col: 'BUREAU_' + col for col in bureau_columns if col != 'SK_ID_CURR'} 
# bureau.rename(columns=new_bureau_columns, inplace=True)

In [4]:
installments = pd.read_csv('../data/dseb63_installment_gb.csv', index_col=0)
installments_columns = installments.columns

installments.drop(columns= [col for col in installments_columns if 'TARGET' in col], inplace=True)

new_installments_columns = {col: 'INSTALLMENTS_' + col for col in installments_columns if col != 'SK_ID_CURR'}
installments.rename(columns=new_installments_columns, inplace=True)

In [5]:
installments.columns

Index(['INSTALLMENTS_NUM_INSTALMENT', 'INSTALLMENTS_NUM_INSTALMENT_NUMBER',
       'INSTALLMENTS_MEAN_INSTALLMENT', 'INSTALLMENTS_LATE_PAYMENT',
       'INSTALLMENTS_EARLY_PAYMENT', 'INSTALLMENTS_LASTEST_PAYMENT_DATE',
       'INSTALLMENTS_TOTAL_PAYMENT', 'INSTALLMENTS_LAST_30_DAYS',
       'INSTALLMENTS_LAST_60_DAYS', 'INSTALLMENTS_LAST_90_DAYS',
       'INSTALLMENTS_LAST_180_DAYS', 'INSTALLMENTS_LAST_365_DAYS',
       'INSTALLMENTS_LAST_730_DAYS', 'INSTALLMENTS_PAYMENT_LAST_30_DAYS',
       'INSTALLMENTS_PAYMENT_LAST_60_DAYS',
       'INSTALLMENTS_PAYMENT_LAST_90_DAYS',
       'INSTALLMENTS_PAYMENT_LAST_180_DAYS',
       'INSTALLMENTS_PAYMENT_LAST_365_DAYS',
       'INSTALLMENTS_PAYMENT_LAST_730_DAYS',
       'INSTALLMENTS_NUM_INSTALMENT_SHORT_TERM',
       'INSTALLMENTS_NUM_INSTALMENT_NUMBER_SHORT_TERM',
       'INSTALLMENTS_MEAN_INSTALLMENT_SHORT_TERM',
       'INSTALLMENTS_LATE_PAYMENT_SHORT_TERM',
       'INSTALLMENTS_EARLY_PAYMENT_SHORT_TERM',
       'INSTALLMENTS_LASTEST_PAYMEN

In [6]:
posh_cash = pd.read_csv('../data/dseb63_pos_cash_gb.csv', index_col=0)
posh_cash_columns = posh_cash.columns

new_posh_cash_columns = {col: 'POSH_CASH_' + col for col in posh_cash_columns if col != 'SK_ID_CURR'}
posh_cash.rename(columns=new_posh_cash_columns, inplace=True)

In [7]:
df = df.merge(bureau, on='SK_ID_CURR', how='left')
df = df.merge(installments, on='SK_ID_CURR', how='left')
df = df.merge(posh_cash, on='SK_ID_CURR', how='left')


In [8]:
df

,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY_x,AMT_GOODS_PRICE,...,POSH_CASH_SHORT_TERM_RATE_COMPLETED,POSH_CASH_SHORT_TERM_SK_ID_PREV,POSH_CASH_SHORT_TERM_NUM_INSTALMENT,POSH_CASH_SHORT_TERM_LAST_12_MONTHS,POSH_CASH_SHORT_TERM_LAST_36_MONTHS,POSH_CASH_NOT_COMPLETED_SHORT_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LONG_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LAST_6_MONTHS,POSH_CASH_NOT_COMPLETED_LAST_12_MONTHS,POSH_CASH_NOT_COMPLETED_CNT_INSTALMENT_FUTURE
0,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,1129500.0,...,1.125000,2.0,16.0,0.0,16.0,0.0,0.0,0.0,0.0,1.0
1,0,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,297000.0,...,1.500000,1.0,8.0,0.0,8.0,0.0,0.0,6.0,10.0,3.0
2,0,Cash loans,M,N,Y,0,121500.0,513000.0,21865.5,513000.0,...,0.948718,3.0,44.0,2.0,31.0,0.0,0.0,6.0,12.0,14.0
3,0,Cash loans,M,N,Y,0,99000.0,490495.5,27517.5,454500.0,...,0.613636,3.0,74.0,11.0,25.0,NaN,NaN,NaN,NaN,NaN
4,0,Cash loans,M,Y,Y,0,360000.0,1530000.0,42075.0,1530000.0,...,0.909091,1.0,11.0,0.0,11.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
246004,0,Cash loans,M,N,N,0,157500.0,254700.0,27558.0,225000.0,...,0.888889,1.0,9.0,9.0,9.0,NaN,NaN,NaN,NaN,NaN
246005,0,Cash loans,F,N,Y,0,72000.0,269550.0,12001.5,225000.0,...,0.857143,1.0,7.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN
246006,0,Cash loans,F,N,Y,0,153000.0,677664.0,29979.0,585000.0,...,1.833333,2.0,10.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
246007,1,Cash loans,F,N,Y,0,171000.0,370107.0,20205.0,319500.0,...,NaN,NaN,NaN,NaN,NaN,0.0,0.0,12.0,20.0,12.0


In [9]:
df.describe()

,TARGET,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY_x,AMT_GOODS_PRICE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,...,POSH_CASH_SHORT_TERM_RATE_COMPLETED,POSH_CASH_SHORT_TERM_SK_ID_PREV,POSH_CASH_SHORT_TERM_NUM_INSTALMENT,POSH_CASH_SHORT_TERM_LAST_12_MONTHS,POSH_CASH_SHORT_TERM_LAST_36_MONTHS,POSH_CASH_NOT_COMPLETED_SHORT_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LONG_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LAST_6_MONTHS,POSH_CASH_NOT_COMPLETED_LAST_12_MONTHS,POSH_CASH_NOT_COMPLETED_CNT_INSTALMENT_FUTURE
count,246009.000000,246009.000000,2.460090e+05,2.460090e+05,245999.000000,2.457850e+05,246009.000000,246009.000000,246009.000000,246009.000000,...,201948.000000,201948.000000,201948.000000,201948.000000,201948.000000,125364.0,125364.0,125364.000000,125364.000000,125364.000000
mean,0.080794,0.416229,1.684589e+05,5.996410e+05,27120.238129,5.389827e+05,0.020855,-16038.946343,63852.612441,-4985.222248,...,1.279040,2.180873,22.539738,2.969641,12.233852,0.0,0.0,4.363374,7.281293,11.473461
std,0.272519,0.720664,1.045297e+05,4.030980e+05,14510.634712,3.700520e+05,0.013818,4361.265483,141306.376949,3521.391396,...,0.829572,1.545564,18.264478,4.206419,11.367394,0.0,0.0,3.131766,5.842464,14.243901
min,0.000000,0.000000,2.565000e+04,4.500000e+04,1615.500000,4.050000e+04,0.000290,-25229.000000,-17912.000000,-23416.000000,...,0.041667,1.000000,1.000000,0.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000
25%,0.000000,0.000000,1.125000e+05,2.700000e+05,16506.000000,2.385000e+05,0.010006,-19682.000000,-2758.000000,-7482.000000,...,0.890110,1.000000,11.000000,0.000000,4.000000,0.0,0.0,2.000000,2.000000,1.000000
50%,0.000000,0.000000,1.462500e+05,5.147775e+05,24903.000000,4.500000e+05,0.018850,-15755.000000,-1212.000000,-4504.000000,...,0.969069,2.000000,17.000000,0.000000,10.000000,0.0,0.0,5.000000,8.000000,6.000000
75%,0.000000,1.000000,2.025000e+05,8.086500e+05,34654.500000,6.795000e+05,0.028663,-12418.000000,-288.000000,-2006.000000,...,1.367521,3.000000,29.000000,6.000000,17.000000,0.0,0.0,6.000000,11.000000,16.000000
max,1.000000,19.000000,1.350000e+07,4.050000e+06,258025.500000,4.050000e+06,0.072508,-7489.000000,365243.000000,0.000000,...,24.000000,21.000000,214.000000,42.000000,111.000000,0.0,0.0,24.000000,48.000000,132.000000


In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 246009 entries, 0 to 246008
Columns: 309 entries, TARGET to POSH_CASH_NOT_COMPLETED_CNT_INSTALMENT_FUTURE
dtypes: float64(252), int64(41), object(16)
memory usage: 580.0+ MB


In [11]:
for col in df.columns:
    print(col, df[col].dtype)

TARGET int64
NAME_CONTRACT_TYPE object
CODE_GENDER object
FLAG_OWN_CAR object
FLAG_OWN_REALTY object
CNT_CHILDREN int64
AMT_INCOME_TOTAL float64
AMT_CREDIT float64
AMT_ANNUITY_x float64
AMT_GOODS_PRICE float64
NAME_TYPE_SUITE object
NAME_INCOME_TYPE object
NAME_EDUCATION_TYPE object
NAME_FAMILY_STATUS object
NAME_HOUSING_TYPE object
REGION_POPULATION_RELATIVE float64
DAYS_BIRTH int64
DAYS_EMPLOYED int64
DAYS_REGISTRATION float64
DAYS_ID_PUBLISH int64
OWN_CAR_AGE float64
FLAG_MOBIL int64
FLAG_EMP_PHONE int64
FLAG_WORK_PHONE int64
FLAG_CONT_MOBILE int64
FLAG_PHONE int64
FLAG_EMAIL int64
OCCUPATION_TYPE object
CNT_FAM_MEMBERS float64
REGION_RATING_CLIENT int64
REGION_RATING_CLIENT_W_CITY int64
WEEKDAY_APPR_PROCESS_START object
HOUR_APPR_PROCESS_START int64
REG_REGION_NOT_LIVE_REGION int64
REG_REGION_NOT_WORK_REGION int64
LIVE_REGION_NOT_WORK_REGION int64
REG_CITY_NOT_LIVE_CITY int64
REG_CITY_NOT_WORK_CITY int64
LIVE_CITY_NOT_WORK_CITY int64
ORGANIZATION_TYPE object
EXT_SOURCE_1 float64
EX

In [12]:
df['EMERGENCYSTATE_MODE'].value_counts()

EMERGENCYSTATE_MODE
No     127434
Yes      1883
Name: count, dtype: int64

In [13]:
df.describe()

,TARGET,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY_x,AMT_GOODS_PRICE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,...,POSH_CASH_SHORT_TERM_RATE_COMPLETED,POSH_CASH_SHORT_TERM_SK_ID_PREV,POSH_CASH_SHORT_TERM_NUM_INSTALMENT,POSH_CASH_SHORT_TERM_LAST_12_MONTHS,POSH_CASH_SHORT_TERM_LAST_36_MONTHS,POSH_CASH_NOT_COMPLETED_SHORT_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LONG_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LAST_6_MONTHS,POSH_CASH_NOT_COMPLETED_LAST_12_MONTHS,POSH_CASH_NOT_COMPLETED_CNT_INSTALMENT_FUTURE
count,246009.000000,246009.000000,2.460090e+05,2.460090e+05,245999.000000,2.457850e+05,246009.000000,246009.000000,246009.000000,246009.000000,...,201948.000000,201948.000000,201948.000000,201948.000000,201948.000000,125364.0,125364.0,125364.000000,125364.000000,125364.000000
mean,0.080794,0.416229,1.684589e+05,5.996410e+05,27120.238129,5.389827e+05,0.020855,-16038.946343,63852.612441,-4985.222248,...,1.279040,2.180873,22.539738,2.969641,12.233852,0.0,0.0,4.363374,7.281293,11.473461
std,0.272519,0.720664,1.045297e+05,4.030980e+05,14510.634712,3.700520e+05,0.013818,4361.265483,141306.376949,3521.391396,...,0.829572,1.545564,18.264478,4.206419,11.367394,0.0,0.0,3.131766,5.842464,14.243901
min,0.000000,0.000000,2.565000e+04,4.500000e+04,1615.500000,4.050000e+04,0.000290,-25229.000000,-17912.000000,-23416.000000,...,0.041667,1.000000,1.000000,0.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000
25%,0.000000,0.000000,1.125000e+05,2.700000e+05,16506.000000,2.385000e+05,0.010006,-19682.000000,-2758.000000,-7482.000000,...,0.890110,1.000000,11.000000,0.000000,4.000000,0.0,0.0,2.000000,2.000000,1.000000
50%,0.000000,0.000000,1.462500e+05,5.147775e+05,24903.000000,4.500000e+05,0.018850,-15755.000000,-1212.000000,-4504.000000,...,0.969069,2.000000,17.000000,0.000000,10.000000,0.0,0.0,5.000000,8.000000,6.000000
75%,0.000000,1.000000,2.025000e+05,8.086500e+05,34654.500000,6.795000e+05,0.028663,-12418.000000,-288.000000,-2006.000000,...,1.367521,3.000000,29.000000,6.000000,17.000000,0.0,0.0,6.000000,11.000000,16.000000
max,1.000000,19.000000,1.350000e+07,4.050000e+06,258025.500000,4.050000e+06,0.072508,-7489.000000,365243.000000,0.000000,...,24.000000,21.000000,214.000000,42.000000,111.000000,0.0,0.0,24.000000,48.000000,132.000000


In [14]:
ordinal_cols = ['FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'EMERGENCYSTATE_MODE']
float_cols = []
int_cols = []
cate_cols = []
for col in df.columns:
    if col not in ordinal_cols:
        if df[col].dtype == 'float64':
            float_cols.append(col)
        elif df[col].dtype == 'int64':
            int_cols.append(col)
        else:
            cate_cols.append(col)


In [15]:
def process_df(df):
    df = df.copy()
    df['EMERGENCYSTATE_MODE'].fillna('No', inplace=True)
    df['FONDKAPREMONT_MODE'].fillna('Unknown', inplace=True)
    df['WALLSMATERIAL_MODE'].fillna('Not Specified', inplace=True)
    df['OCCUPATION_TYPE'].replace('IT staff', 'High skill tech staff', inplace=True)
    df['OCCUPATION_TYPE'].replace('Realty agents', 'Sales staff', inplace=True)
    df['OCCUPATION_TYPE'].replace('HR staff', 'Laborers', inplace=True)
    df['OCCUPATION_TYPE'].fillna('Unknown', inplace=True)
    df['NAME_EDUCATION_TYPE'].replace('Academic degree', 'Higher education', inplace=True)

    others = df['NAME_INCOME_TYPE'].value_counts().index[4:]
    df['NAME_INCOME_TYPE'].replace(others, 'Others', inplace=True)
    df['NAME_TYPE_SUITE'].fillna('Unaccompanied', inplace=True)

    df['OWN_CAR_AGE'].fillna(-100, inplace=True)
    return df
df = process_df(df)

/tmp/ipykernel_140881/112243189.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['EMERGENCYSTATE_MODE'].fillna('No', inplace=True)
/tmp/ipykernel_140881/112243189.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)'

In [16]:
def display_missing_data_info(dataframe, ascending=False):
    missing_values_count = dataframe.isnull().sum()
    missing_values_percentage = (dataframe.isnull().mean() * 100)
    missing_data = pd.DataFrame({
        'Missing Values': missing_values_count,
        'Percentage (%)': missing_values_percentage
    })
    
    missing_data = missing_data[missing_data['Missing Values'] > 0]
    
    sorted_missing_data = missing_data.sort_values(by='Missing Values', ascending=ascending)

    print(sorted_missing_data)
    return sorted_missing_data

In [17]:
display_missing_data_info(df)

                        Missing Values  Percentage (%)
BAD_DEBT_DURATION               245993       99.993496
COUNT_BAD_DEBT                  245993       99.993496
BAD_DEBT_SUM_DEBT               245993       99.993496
BAD_DEBT_SUM_CREDIT             245993       99.993496
BAD_DEBT_FINISHED               245993       99.993496
...                                ...             ...
EXT_SOURCE_2                       529        0.215033
AMT_GOODS_PRICE                    224        0.091054
AMT_ANNUITY_x                       10        0.004065
DAYS_LAST_PHONE_CHANGE               1        0.000406
CNT_FAM_MEMBERS                      1        0.000406

[248 rows x 2 columns]


,Missing Values,Percentage (%)
BAD_DEBT_DURATION,245993,99.993496
COUNT_BAD_DEBT,245993,99.993496
BAD_DEBT_SUM_DEBT,245993,99.993496
BAD_DEBT_SUM_CREDIT,245993,99.993496
BAD_DEBT_FINISHED,245993,99.993496
...,...,...
EXT_SOURCE_2,529,0.215033
AMT_GOODS_PRICE,224,0.091054
AMT_ANNUITY_x,10,0.004065
DAYS_LAST_PHONE_CHANGE,1,0.000406


In [18]:
from sklearn.preprocessing import OneHotEncoder

from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler, PowerTransformer, OrdinalEncoder, LabelEncoder
from sklearn.impute import SimpleImputer

In [19]:
df_train = df.copy()
df_train.dropna(subset=['TARGET'], inplace=True)

In [20]:
df_train['TARGET'].value_counts()

TARGET
0    226133
1     19876
Name: count, dtype: int64

In [21]:
X = df_train.drop(columns=['TARGET', 'SK_ID_CURR'])
y = df_train['TARGET']

In [22]:
ordinal_cols = ['FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'EMERGENCYSTATE_MODE']
float_cols = []
int_cols = []
cate_cols = []
flag_cols = []
for col in X.columns:
    if col not in ordinal_cols:
        if X[col].dtype == 'float64':
            float_cols.append(col)
        elif X[col].dtype == 'int64':
            if 'FLAG' in col:
                flag_cols.append(col)
            else:
                int_cols.append(col)
        else:
            cate_cols.append(col)


In [23]:
keep_flag = flag_cols.copy()
# keep_flag = []
# for col in flag_cols:
#     if 'DOCUMENT' not in col:
#         keep_flag.append(col)

In [24]:
for col in float_cols:
    X[col].fillna(0, inplace=True)
    
for col in int_cols:
    X[col].fillna(0, inplace=True)
for col in cate_cols:
    X[col].fillna('Unknown', inplace=True)

/tmp/ipykernel_140881/1735439884.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  X[col].fillna(0, inplace=True)
/tmp/ipykernel_140881/1735439884.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.meth

In [25]:
# Fill missing values in categorical columns


In [26]:
from sklearn.model_selection import train_test_split

X_train_, X_val, y_train, y_val = train_test_split(
    X, y, test_size=0.1, random_state=42, stratify=y
)

In [27]:
y_train = y_train.to_numpy()
y_val = y_val.to_numpy()

In [28]:
onehot_transformer = OneHotEncoder(handle_unknown='ignore')
power_transformer = PowerTransformer()
ordinal_transformer = OrdinalEncoder()
scaler_transformer = StandardScaler()

preprocessor = ColumnTransformer(
    transformers=[
        ('onehot', onehot_transformer, cate_cols),
        ('power', power_transformer, float_cols), # power_transformer
        ('ordinal', ordinal_transformer, ordinal_cols),
        ('scale', power_transformer, int_cols),
        ('flag', 'passthrough', keep_flag)
        
    ]
)


In [29]:
X.head()

,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY_x,AMT_GOODS_PRICE,NAME_TYPE_SUITE,...,POSH_CASH_SHORT_TERM_RATE_COMPLETED,POSH_CASH_SHORT_TERM_SK_ID_PREV,POSH_CASH_SHORT_TERM_NUM_INSTALMENT,POSH_CASH_SHORT_TERM_LAST_12_MONTHS,POSH_CASH_SHORT_TERM_LAST_36_MONTHS,POSH_CASH_NOT_COMPLETED_SHORT_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LONG_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LAST_6_MONTHS,POSH_CASH_NOT_COMPLETED_LAST_12_MONTHS,POSH_CASH_NOT_COMPLETED_CNT_INSTALMENT_FUTURE
0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,1129500.0,Family,...,1.125000,2.0,16.0,0.0,16.0,0.0,0.0,0.0,0.0,1.0
1,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,297000.0,Unaccompanied,...,1.500000,1.0,8.0,0.0,8.0,0.0,0.0,6.0,10.0,3.0
2,Cash loans,M,N,Y,0,121500.0,513000.0,21865.5,513000.0,Unaccompanied,...,0.948718,3.0,44.0,2.0,31.0,0.0,0.0,6.0,12.0,14.0
3,Cash loans,M,N,Y,0,99000.0,490495.5,27517.5,454500.0,"Spouse, partner",...,0.613636,3.0,74.0,11.0,25.0,0.0,0.0,0.0,0.0,0.0
4,Cash loans,M,Y,Y,0,360000.0,1530000.0,42075.0,1530000.0,Unaccompanied,...,0.909091,1.0,11.0,0.0,11.0,0.0,0.0,0.0,0.0,0.0


In [30]:
X_train = preprocessor.fit_transform(X_train_)
X_train

array([[0., 1., 0., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 0.],
       ...,
       [1., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.]])

In [31]:
X_val = preprocessor.transform(X_val)

In [32]:
X_train.shape   

(221408, 425)

In [33]:
cat_feature = preprocessor.named_transformers_['onehot'].get_feature_names_out(cate_cols)
cat_feature

feature_names = np.concatenate([cat_feature, float_cols, ordinal_cols, int_cols, keep_flag])

In [34]:
df_train_preprocessed = pd.DataFrame(X_train, columns=feature_names)
df_train_preprocessed

,NAME_CONTRACT_TYPE_Cash loans,NAME_CONTRACT_TYPE_Revolving loans,CODE_GENDER_F,CODE_GENDER_M,CODE_GENDER_XNA,NAME_TYPE_SUITE_Children,NAME_TYPE_SUITE_Family,NAME_TYPE_SUITE_Group of people,NAME_TYPE_SUITE_Other_A,NAME_TYPE_SUITE_Other_B,...,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21
0,0.0,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,1.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
221403,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
221404,1.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
221405,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
221406,1.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [35]:
for column in df_train_preprocessed.columns:
    print(column, df_train_preprocessed[column].dtype)

NAME_CONTRACT_TYPE_Cash loans float64
NAME_CONTRACT_TYPE_Revolving loans float64
CODE_GENDER_F float64
CODE_GENDER_M float64
CODE_GENDER_XNA float64
NAME_TYPE_SUITE_Children float64
NAME_TYPE_SUITE_Family float64
NAME_TYPE_SUITE_Group of people float64
NAME_TYPE_SUITE_Other_A float64
NAME_TYPE_SUITE_Other_B float64
NAME_TYPE_SUITE_Spouse, partner float64
NAME_TYPE_SUITE_Unaccompanied float64
NAME_INCOME_TYPE_Commercial associate float64
NAME_INCOME_TYPE_Others float64
NAME_INCOME_TYPE_Pensioner float64
NAME_INCOME_TYPE_State servant float64
NAME_INCOME_TYPE_Working float64
NAME_EDUCATION_TYPE_Higher education float64
NAME_EDUCATION_TYPE_Incomplete higher float64
NAME_EDUCATION_TYPE_Lower secondary float64
NAME_EDUCATION_TYPE_Secondary / secondary special float64
NAME_FAMILY_STATUS_Civil marriage float64
NAME_FAMILY_STATUS_Married float64
NAME_FAMILY_STATUS_Separated float64
NAME_FAMILY_STATUS_Single / not married float64
NAME_FAMILY_STATUS_Unknown float64
NAME_FAMILY_STATUS_Widow float

In [36]:
display_missing_data_info(df_train_preprocessed)

Empty DataFrame
Columns: [Missing Values, Percentage (%)]
Index: []


,Missing Values,Percentage (%)


In [37]:
!pip install imbalanced-learn

## Upsample - Downsample

In [38]:
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import ClusterCentroids, TomekLinks
from sklearn.model_selection import train_test_split

In [92]:
smote = SMOTE(sampling_strategy=0.2, random_state=42, n_jobs=-1)
X_resampled, y_resampled = smote.fit_resample(X_train, y_train)

/home/quanghung20gg/anaconda3/lib/python3.12/site-packages/imblearn/over_sampling/_smote/base.py:370: FutureWarning: The parameter `n_jobs` has been deprecated in 0.10 and will be removed in 0.12. You can pass an nearest neighbors estimator where `n_jobs` is already set instead.
  warnings.warn(


In [40]:
# undersampler = TomekLinks(n_jobs=-1)
# X_resampled, y_resampled = smote.fit_resample(X_train, y_train)

### Manual double

In [41]:
from sklearn.utils import resample


In [42]:
y

0         0
1         0
2         0
3         0
4         0
         ..
246004    0
246005    0
246006    0
246007    1
246008    0
Name: TARGET, Length: 246009, dtype: int64

In [43]:
y_train

array([0, 0, 0, ..., 0, 0, 0])

In [96]:
posidx = (y_train == 1)
negidx = (y_train == 0)


# X_train_pos = X_train[posidx]
# y_train_pos = y_train[posidx]

# X_train_neg = X_train[negidx]
# y_train_neg = y_train[negidx]

# X_resampled = np.concatenate([X_train_pos, X_train_neg, X_train_neg])
# y_resampled = np.concatenate([y_train_pos, y_train_neg, y_train_neg])

y_train2 = y_train.reshape(-1, 1)

train_data = np.concatenate([X_train, y_train2], axis=1)

train_data_pos = train_data[posidx]
train_data_neg = train_data[negidx]

train_data_resampled = resample(train_data_pos, n_samples=int(len(train_data_neg)/4), random_state=42)

In [97]:
train_data_resampled = np.concatenate([train_data_resampled, train_data_neg])

In [98]:
np.random.shuffle(train_data_resampled)

In [99]:
train_data_resampled

array([[1., 0., 1., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 1.],
       ...,
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 1., ..., 0., 0., 0.],
       [1., 0., 1., ..., 0., 0., 1.]])

In [100]:
X_resampled = train_data_resampled[:, :-1]
y_resampled = train_data_resampled[:, -1]

In [101]:
X_resampled.shape

(254400, 425)

In [87]:
X_resampled = X_train
y_resampled = y_train

In [88]:
X_resampled.shape

(221408, 425)

In [52]:

from sklearn.model_selection import cross_val_score, StratifiedKFold, cross_validate
from sklearn.ensemble import StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split

In [53]:
from sklearn.decomposition import PCA
import pandas as pd
import numpy as np

In [54]:
pca = PCA()
pca.fit(X_train)

PCA()

In [55]:
explained_variance = pca.explained_variance_ratio_

# Get the absolute values of PCA components (loadings)
feature_contributions = np.abs(pca.components_)

# Aggregate feature importance scores
# Weight each feature's contribution by the variance explained by the component
weighted_contributions = feature_contributions * explained_variance[:, np.newaxis]

# Sum weighted contributions across components
feature_importance = weighted_contributions.sum(axis=0)

# Create a DataFrame for ranking
importance_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': feature_importance
}).sort_values(by='Importance', ascending=False)
importance_df

,Feature,Importance
300,INSTALLMENTS_LAST_30_DAYS,4.319810e-02
306,INSTALLMENTS_PAYMENT_LAST_30_DAYS,4.312927e-02
363,POSH_CASH_LAST_12_MONTHS,4.216791e-02
325,INSTALLMENTS_PAYMENT_LAST_30_DAYS_SHORT_TERM,4.210030e-02
319,INSTALLMENTS_LAST_30_DAYS_SHORT_TERM,4.209825e-02
...,...,...
278,BAD_DEBT_LAST_365_DAYS,1.740203e-19
279,BAD_DEBT_LAST_730_DAYS,1.352994e-19
378,POSH_CASH_NOT_COMPLETED_SHORT_CNT_INSTALMENT,2.561478e-21
379,POSH_CASH_NOT_COMPLETED_LONG_CNT_INSTALMENT,1.636252e-21


In [56]:
def select_features_with_pca(X, threshold=0.1):
    """
    Select features contributing to principal components above a certain threshold.

    Parameters:
    X (numpy array or pandas DataFrame): Input feature matrix.
    threshold (float): Contribution threshold for selecting features (default is 0.1).

    Returns:
    selected_columns (list): List of selected feature names or indices.
    selected_data (DataFrame): Data with selected features.
    """
    # Convert to DataFrame if X is a NumPy array
    # if isinstance(X, np.ndarray):
    #     X = pd.DataFrame(X, columns=[f"Feature{i+1}" for i in range(X.shape[1])])
    
    # Apply PCA
    pca = PCA()
    pca.fit(X)
    
    # Calculate absolute loadings (contributions of features to components)
    explained_variance = pca.explained_variance_ratio_
    feature_contributions = np.abs(pca.components_)
    weighted_contributions = feature_contributions * explained_variance[:, np.newaxis]
    # Identify features exceeding the threshold for any component
    
    feature_mask = (weighted_contributions > threshold).any(axis=0)
    
    # Filter the data to keep only selected features
    # selected_data = X[feature_mask, :]
    
    return feature_mask

IF USING PCA

In [94]:
feature_map = select_features_with_pca(X_train, threshold=0.0001)

In [90]:
feature_map = np.ones(len(feature_names), dtype=bool)

In [114]:
feature_map.sum()

328

In [60]:
# X_df = pd.DataFrame(X_train, columns=feature_names)
# X_train_selected_columns, X_train_selected = select_features_with_pca(X_df, threshold=0.0001)

In [85]:
logistic_model = LogisticRegression(random_state=42, max_iter=10000, n_jobs=-1, class_weight='balanced')

In [62]:
dt_model = DecisionTreeClassifier(random_state=42, class_weight = 'balanced')

In [63]:
stacking_model = StackingClassifier(
    estimators=[
        ('lr', logistic_model),
        ('dt', dt_model)
    ],
    final_estimator=LogisticRegression(class_weight='balanced', random_state=42, max_iter=1000),  # Final model for stacking
    cv=4
)

In [64]:
kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=42)

In [65]:
from sklearn.metrics import make_scorer, roc_auc_score
def gini_coefficient(y_true, y_pred):
    """
    Calculate the Gini coefficient using predictions and true labels.
    
    Parameters:
    y_true (array-like): True binary labels.
    y_pred (array-like): Predicted probabilities.
    
    Returns:
    float: Gini coefficient.
    """
    auc = roc_auc_score(y_true, y_pred)  # AUC calculation
    return 2 * auc - 1  # Gini coefficient

In [66]:
gini_scorer = make_scorer(gini_coefficient, needs_proba=True)

/home/quanghung20gg/anaconda3/lib/python3.12/site-packages/sklearn/metrics/_scorer.py:548: FutureWarning: The `needs_threshold` and `needs_proba` parameter are deprecated in version 1.4 and will be removed in 1.6. You can either let `response_method` be `None` or set it to `predict` to preserve the same behaviour.
  warnings.warn(


In [102]:
print("Logistic Regression Cross-Validation Scores:")
results = cross_validate(logistic_model, X_resampled[:, feature_map], y_resampled, cv=kfold, scoring=gini_scorer, n_jobs=-1,return_estimator=True)

for i, estimator in enumerate(results['estimator']):
    y_pred = estimator.predict_proba(X_val[:, feature_map])
    gini_scores = gini_coefficient(y_val, y_pred[:,1])
    print(f"- Gini Score for fold {i}: {gini_scores:.4f}")

Logistic Regression Cross-Validation Scores:
- Gini Score for fold 0: 0.5156
- Gini Score for fold 1: 0.5158
- Gini Score for fold 2: 0.5183
- Gini Score for fold 3: 0.5186


### RAW
Logistic Regression Cross-Validation Scores:
- Gini Score for fold 0: 0.5146
- Gini Score for fold 1: 0.5154
- Gini Score for fold 2: 0.5152
- Gini Score for fold 3: 0.5153

### UPSAMPLE RAW

Logistic Regression Cross-Validation Scores:
- Gini Score for fold 0: 0.5135
- Gini Score for fold 1: 0.5167
- Gini Score for fold 2: 0.5118
- Gini Score for fold 3: 0.5127

### PCA
Logistic Regression Cross-Validation Scores:
- Gini Score for fold 0: 0.5148
- Gini Score for fold 1: 0.5187
- Gini Score for fold 2: 0.5145
- Gini Score for fold 3: 0.5136



### Drop Flag

Logistic Regression Cross-Validation Scores:
- Gini Score for fold 0: 0.5129
- Gini Score for fold 1: 0.5156
- Gini Score for fold 2: 0.5125
- Gini Score for fold 3: 0.5151

### Super upsample + PCA
- Gini Score for fold 0: 0.5156
- Gini Score for fold 1: 0.5158
- Gini Score for fold 2: 0.5183
- Gini Score for fold 3: 0.5186

In [216]:
print("Decision Tree Cross-Validation Scores:")
results = cross_validate(dt_model, X_resampled[:, feature_map], y_resampled, cv=kfold, scoring=gini_scorer, n_jobs=-1,return_estimator=True)

for i, estimator in enumerate(results['estimator']):
    y_pred = estimator.predict_proba(X_val)
    gini_scores = gini_coefficient(y_val, y_pred[:,1])
    print(f"Gini Score for fold {i}: {gini_scores:.4f}")

Decision Tree Cross-Validation Scores:


KeyboardInterrupt: 

In [177]:
print("Stacking Scores:")
stacking_model.fit(X_resampled[:, feature_map], y_resampled)
y_pred = stacking_model.predict_proba(X_val[:, feature_map])
gini_scores = gini_coefficient(y_val, y_pred[:,1])
print(f"Gini Scores: {gini_scores}")

Stacking Scores:
Gini Scores: 0.4890627984953009


In [ ]:
# print("Stacking Cross-Validation Scores:")
# results = cross_validate(stacking_model, X_resampled, y_resampled, cv=4, scoring=gini_scorer, n_jobs=-1,return_estimator=True)
# print(f"Gini Scores for each fold: {results['test_score']}")
# print(f"Mean Gini Score: {results['test_score'].mean():.4f}")

In [ ]:
X_resampled.shape

(223872, 428)

In [ ]:
# for i, estimator in enumerate(results['estimator']):
#     y_pred = estimator.predict_proba(X_val)
#     gini_scores = gini_coefficient(y_val, y_pred[:,1])
#     print(f"Gini Score for fold {i}: {gini_scores:.4f}")

NameError: name 'results' is not defined

Stacking Cross-Validation Scores:

Gini Scores for each fold: [0.45493243 0.47045351 0.46576538 0.47327105 0.45963799]

Mean Gini Score: 0.4648

In [103]:
df_test = pd.read_csv('../data/dseb63_application_test.csv',  index_col=0)


In [104]:
df_test = df_test.merge(bureau, on='SK_ID_CURR', how='left')
df_test = df_test.merge(installments, on='SK_ID_CURR', how='left')
df_test = df_test.merge(posh_cash, on='SK_ID_CURR', how='left')

In [105]:
df_test = process_df(df_test)

/tmp/ipykernel_140881/112243189.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['EMERGENCYSTATE_MODE'].fillna('No', inplace=True)
/tmp/ipykernel_140881/112243189.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)'

In [106]:
df_test 

,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY_x,AMT_GOODS_PRICE,NAME_TYPE_SUITE,...,POSH_CASH_SHORT_TERM_RATE_COMPLETED,POSH_CASH_SHORT_TERM_SK_ID_PREV,POSH_CASH_SHORT_TERM_NUM_INSTALMENT,POSH_CASH_SHORT_TERM_LAST_12_MONTHS,POSH_CASH_SHORT_TERM_LAST_36_MONTHS,POSH_CASH_NOT_COMPLETED_SHORT_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LONG_CNT_INSTALMENT,POSH_CASH_NOT_COMPLETED_LAST_6_MONTHS,POSH_CASH_NOT_COMPLETED_LAST_12_MONTHS,POSH_CASH_NOT_COMPLETED_CNT_INSTALMENT_FUTURE
0,Cash loans,M,Y,N,2,207000.0,465457.5,52641.0,418500.0,Unaccompanied,...,0.923077,1.0,13.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0
1,Cash loans,F,Y,Y,0,247500.0,1281712.5,48946.5,1179000.0,Unaccompanied,...,2.368603,7.0,40.0,0.0,22.0,0.0,0.0,0.0,0.0,1.0
2,Cash loans,F,Y,N,0,202500.0,495000.0,39109.5,495000.0,Unaccompanied,...,0.854545,2.0,21.0,8.0,11.0,NaN,NaN,NaN,NaN,NaN
3,Cash loans,F,N,Y,0,247500.0,254700.0,24939.0,225000.0,Unaccompanied,...,0.480000,1.0,25.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN
4,Cash loans,M,N,Y,0,112500.0,308133.0,15862.5,234000.0,Unaccompanied,...,2.147436,4.0,26.0,3.0,7.0,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
61497,Cash loans,M,Y,Y,0,171000.0,521280.0,23089.5,450000.0,Family,...,1.427751,7.0,90.0,7.0,43.0,0.0,0.0,10.0,22.0,3.0
61498,Revolving loans,M,Y,Y,2,450000.0,900000.0,45000.0,900000.0,Family,...,0.872381,3.0,44.0,0.0,14.0,0.0,0.0,6.0,10.0,1.0
61499,Cash loans,F,N,Y,0,225000.0,202500.0,24030.0,202500.0,Family,...,3.677211,7.0,38.0,5.0,21.0,0.0,0.0,0.0,0.0,1.0
61500,Cash loans,M,N,Y,0,121500.0,254700.0,30357.0,225000.0,Family,...,0.900000,2.0,21.0,0.0,0.0,NaN,NaN,NaN,NaN,NaN


In [107]:
for col in float_cols:
    df_test[col].fillna(0, inplace=True)
    
for col in int_cols:
    df_test[col].fillna(0, inplace=True)
for col in cate_cols:
    df_test[col].fillna('Unknown', inplace=True)

/tmp/ipykernel_140881/1295519593.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_test[col].fillna(0, inplace=True)
/tmp/ipykernel_140881/1295519593.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'd

In [108]:
X_test = df_test.drop(columns=['SK_ID_CURR'])

In [109]:
X_test = preprocessor.transform(X_test)


In [110]:
X_all = np.concatenate([X_resampled, X_val])
y_all = np.concatenate([y_resampled, y_val])

In [115]:
logistic_model.fit(X_all[:, feature_map], y_all)

LogisticRegression(class_weight='balanced', max_iter=10000, n_jobs=-1,
                   random_state=42)

In [117]:
y_pred = logistic_model.predict_proba(X_test[:, feature_map])   

In [ ]:
stacking_model.fit(X_all[:, feature_map], y_all)

StackingClassifier(cv=4,
                   estimators=[('lr',
                                LogisticRegression(class_weight='balanced',
                                                   max_iter=10000, n_jobs=-1,
                                                   random_state=42)),
                               ('dt',
                                DecisionTreeClassifier(class_weight='balanced',
                                                       random_state=42))],
                   final_estimator=LogisticRegression(class_weight='balanced',
                                                      max_iter=1000,
                                                      random_state=42))

In [ ]:
y_pred = stacking_model.predict_proba(X_test[:, feature_map])

In [ ]:
# gini_coefficient(y, y_pred[:, 1])

In [118]:
submission = pd.DataFrame({
    'SK_ID_CURR': df_test['SK_ID_CURR'],
    'TARGET': y_pred[:, 1]
})
submission.to_csv('../submission/dseb63_log_power_pca_bureau_installment_pos_cash.csv', index=False)